text/3446-store.md

```diff
@@ -0,0 +1,998 @@
+            //  -   `pointer` is properly aligned.
+            unsafe { ptr::drop_in_place(pointer) }
+        };
+
+        let layout = if self.is_left {
+            Layout::new::<L>()
+        } else {
+            Layout::new::<R>()
+        };
+
+        //  Safety:
+        //  -   `self.store` will no longer be used.
+        let store = unsafe { ManuallyDrop::take(&mut self.store) };
+
+        //  Safety:
+        //  -   `self.handle` was allocated by `self.store`.
+        //  -   `self.handle` is still valid.
+        //  -   `layout` fits the block of memory associated to `self.handle`.
+        unsafe { store.deallocate(self.handle, layout) }
+    }
+}
+```
+
+By using `Store`, I empower my users to use my type in a wide variety of scenarios, some of which I cannot even fathom.
+
+The overhead of using `Store` over `Allocator` is also fairly low, so that the added flexibility comes at little to no
+cost to me.
+
+More examples of collections can be found at https://github.com/matthieu-m/storage/tree/main/src/collection, including
+a complete linked list, a box draft, a concurrent vector draft, and a skip list draft.
+
+
+##  Store Implementer Guide
+
+The API of `Store` requires internal mutability, and that's it. I can otherwise provide as few or as many guarantees as
+I wish.
+
+```rust
+/// An implementation of `Store` providing a single, in-line, block of memory.
+///
+/// The block of memory is aligned and sized as per `T`.
+pub struct InlineSingleStore<T>(UnsafeCell<MaybeUninit<T>>);
+
+impl<T> Default for InlineSingleStore<T> {
+    fn default() -> Self {
+        Self(UnsafeCell::new(MaybeUninit::uninit()))
+    }
+}
+
+unsafe impl<T> Store for InlineSingleStore<T> {
+    type Handle = ();
+
+    fn dangling(&self) -> Self::Handle {}
+
+    fn allocate(&self, layout: Layout) -> Result<(Self::Handle, usize), AllocError> {
+        Self::validate_layout(layout)?;
+
+        Ok(((), mem::size_of::<T>()))
+    }
+
+    unsafe fn deallocate(&self, _handle: Self::Handle, _layout: Layout) {}
+
+    unsafe fn resolve(&self, _handle: Self::Handle) -> NonNull<u8> {
+        let pointer = self.0.get();
+
+        //  Safety:
+        //  -   `self` is non null.
+        unsafe { NonNull::new_unchecked(pointer) }.cast()
+    }
+
+    unsafe fn grow(
+        &self,
+        _handle: Self::Handle,
+        _old_layout: Layout,
+        new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError> {
+        debug_assert!(
+            new_layout.size() >= _old_layout.size(),
+            "{new_layout:?} must have a greater size than {_old_layout:?}"
+        );
+
+        Self::validate_layout(new_layout)?;
+
+        Ok(((), mem::size_of::<T>()))
+    }
+
+    unsafe fn shrink(
+        &self,
+        _handle: Self::Handle,
+        _old_layout: Layout,
+        _new_layout: Layout,
+    ) -> Result<(Self::Handle, usize), AllocError> {
+        debug_assert!(
+            _new_layout.size() >= _old_layout.size(),
```

Comment:
```suggestion
            _new_layout.size() <= _old_layout.size(),
```

Reply:
Oops, I'll push that in the batch of fixes.